### PR TITLE
catalog: Move catalog data migrations to the catalog crate

### PR DESCRIPTION
**Pull request by jkosh44**
This PR moves the code for catalog data migrations from the stash crate, and other crates, to the catalog crate. This will help us port the migrations over to a more generic format that will work with the persist catalog.

### Motivation
This PR refactors existing code.

### Tips for reviewer
* Reviewing each commit separately may be easier the reviewing all at once.
* This PR is almost completely code movement, but it's a lot of code movement. The contents of every single `impl RustType<_> for _` are unchanged.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/catalog/src/durable/objects/serialization.rs
Comment (jkosh44) [inline src/catalog/src/durable/objects/serialization.rs:1]: This file is a collection of all the `RustType` impls and is pure code movement.

**Comment by jkosh44** [comment]
@benesch you were auto-added due to changes in workspace-hack. Feel free to remove yourself (or look at the workspace-hack changes if you want).

**Review by ParkMyCar — COMMENTED** [review]
What's left in the `stash-types` crate after this change? Does this regress our compile times by moving the protos into the `catalog` crate? Does it make sense to move the protobuf files into a (or rename the `stash-types`) `catalog-types` crate?

**Comment by jkosh44** [comment]
> What's left in the `stash-types` crate after this change?

`StashError` and `mz_stash_types::Metrics` are left. It's probably worth folding those back into `mz_stash`.

> Does this regress our compile times by moving the protos into the `catalog` crate?

I'm not an expert on this, but my thinking is that compiling from zero is unchanged. However, compiling from an `mz_catalog` invalidation has gotten longer.

> Does it make sense to move the protobuf files into a (or rename the `stash-types`) `catalog-types` crate?

I think a `catalog-types` might be helpful here. I'll whip something up.

**Comment by jkosh44** [comment]
@ParkMyCar, just wanted to summarize what we spoke about in person.
- `clusterd` still depends on `mz_stash_types` indirectly, and it's still a nice property that we want to maintain that changing `mz_stash` doesn't recompile `clusterd`. So we're leaving `mz_stash_types` as is.
- `mz_catalog` was not added as a dependency to any other crates and the protos are only re-built when they are changed. So, compile times should not be affected by this PR.

**Review by ParkMyCar — APPROVED** [review]
Like Joe mentioned above, we chatted in person and this is good to go

**Pull request merged**